Basic VDO creation role

GregSutcliffe opened the pull request:
This works for creating a VDO-backed device. I'm opening this one first to check I'm getting the basic usage of our structure right - I'll do another PR for Zabbix monitoring and for idempotency.

GregSutcliffe committed:
Basic VDO creation role

GregSutcliffe committed:
Add fs creation / management

GregSutcliffe commented:
Added filesystem creation. Tested idempotence of the VG and the FS:

```
[root@n27-0-253 test]# ansible-playbook main.yaml -i inventory.yml 

PLAY [localhost] ********************************************************************************************************************************************

TASK [Gathering Facts] **************************************************************************************************************************************
ok: [localhost]

TASK [vdo : Importing specific distro variables] ************************************************************************************************************
ok: [localhost] => (item=/root/.ansible/roles/vdo/vars/CentOS.yml)

TASK [vdo : Install the VDO packages] ***********************************************************************************************************************
ok: [localhost]

TASK [vdo : Create the volume group] ************************************************************************************************************************
ok: [localhost]

TASK [vdo : Setup VDO storage pool] *************************************************************************************************************************
changed: [localhost]

TASK [vdo : Create vdo mount point] *************************************************************************************************************************
ok: [localhost]

TASK [vdo : Format the xfs filesystem] **********************************************************************************************************************
changed: [localhost]

TASK [vdo : Mount the lv] ***********************************************************************************************************************************
changed: [localhost]

PLAY RECAP **************************************************************************************************************************************************
localhost                  : ok=8    changed=3    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   
```

GregSutcliffe committed:
Add idempotence check and manage LVs if present

arrfab commented on tasks/main.yml:
```diff
@@ -8,3 +9,36 @@
+    mode: '0755'
+- name: Format the xfs filesystem
+  community.general.filesystem:
+    fstype: xfs
```
That should be a variable as by default we use ext4, so at least make that one a variable too ?

GregSutcliffe replied:
Added

arrfab commented on tasks/main.yml:
```diff
@@ -8,3 +9,36 @@
+  ansible.posix.mount:
+    path: "{{ vdo_mount_point }}"
+    src: "/dev/{{ vdo_vg_name }}/{{ vdo_lv_name }}"
+    fstype: xfs
```
same remark about fstype (should be a variable)

GregSutcliffe replied:
Added

arrfab commented on tasks/main.yml:
```diff
@@ -8,3 +9,36 @@
+  community.general.lvol:
+    vg: "{{ vdo_vg_name }}"
+    lv: "{{ vdo_lv_name }}"
+    size: "{{ vdo_phys_size }}"
```
Can we have also a subsequent lvol run to set the lv size accordingly ? so if we want to increase underlying size we can also reflect a new virtualsize on the vdo lv itself

GregSutcliffe replied:
Yep, I'm just adding that now. Will do the FS type too.

GregSutcliffe commented:
Added idempotence and management of both the physical `vpool` and the thin LV. Example runs:

Defaults with no LV present, but VG already created:
```
[root@n27-0-253 test]# ansible-playbook main.yaml -i inventory.yml 

PLAY [localhost] ********************************************************************************************************************************************

TASK [Gathering Facts] **************************************************************************************************************************************
ok: [localhost]

TASK [vdo : Importing specific distro variables] ************************************************************************************************************
ok: [localhost] => (item=/root/.ansible/roles/vdo/vars/CentOS.yml)

TASK [vdo : Install the VDO packages] ***********************************************************************************************************************
ok: [localhost]

TASK [vdo : Create the volume group] ************************************************************************************************************************
ok: [localhost]

TASK [vdo : Check if the path exists] ***********************************************************************************************************************
ok: [localhost]

TASK [vdo : Setup VDO pool & LV] ****************************************************************************************************************************
changed: [localhost]

TASK [vdo : Manage VDO storage pool] ************************************************************************************************************************
skipping: [localhost]

TASK [vdo : Manage VDO thin LV] *****************************************************************************************************************************
skipping: [localhost]

TASK [vdo : Create vdo mount point] *************************************************************************************************************************
ok: [localhost]

TASK [vdo : Format the xfs filesystem] **********************************************************************************************************************
changed: [localhost]

TASK [vdo : Mount the lv] ***********************************************************************************************************************************
changed: [localhost]

PLAY RECAP **************************************************************************************************************************************************
localhost                  : ok=9    changed=3    unreachable=0    failed=0    skipped=2    rescued=0    ignored=0   

[root@n27-0-253 test]# lvs
  LV     VG    Attr       LSize   Pool   Origin Data%  Meta%  Move Log Cpy%Sync Convert
  vdolv  vdovg vwi-aov--- 100.00g vpool0        0.01                                   
  vpool0 vdovg dwi-------  20.00g               20.06                                  

[root@n27-0-253 test]# df -h /mnt/test/ 
Filesystem               Size  Used Avail Use% Mounted on
/dev/mapper/vdovg-vdolv  100G  746M  100G   1% /mnt/test
```
Now re-run for idempotence:
```
[root@n27-0-253 test]# ansible-playbook main.yaml -i inventory.yml 

PLAY [localhost] ********************************************************************************************************************************************

TASK [Gathering Facts] **************************************************************************************************************************************
ok: [localhost]

TASK [vdo : Importing specific distro variables] ************************************************************************************************************
ok: [localhost] => (item=/root/.ansible/roles/vdo/vars/CentOS.yml)

TASK [vdo : Install the VDO packages] ***********************************************************************************************************************
ok: [localhost]

TASK [vdo : Create the volume group] ************************************************************************************************************************
ok: [localhost]

TASK [vdo : Check if the path exists] ***********************************************************************************************************************
ok: [localhost]

TASK [vdo : Setup VDO pool & LV] ****************************************************************************************************************************
skipping: [localhost]

TASK [vdo : Manage VDO storage pool] ************************************************************************************************************************
changed: [localhost]

TASK [vdo : Manage VDO thin LV] *****************************************************************************************************************************
ok: [localhost]

TASK [vdo : Create vdo mount point] *************************************************************************************************************************
ok: [localhost]

TASK [vdo : Format the xfs filesystem] **********************************************************************************************************************
ok: [localhost]

TASK [vdo : Mount the lv] ***********************************************************************************************************************************
ok: [localhost]

PLAY RECAP **************************************************************************************************************************************************
localhost                  : ok=10   changed=1    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0   
```

Now add 10G to the `vpool`:
```
[root@n27-0-253 test]# vi inventory.yml 
[root@n27-0-253 test]# ansible-playbook main.yaml -i inventory.yml 

PLAY [localhost] ********************************************************************************************************************************************

TASK [Gathering Facts] **************************************************************************************************************************************
ok: [localhost]

TASK [vdo : Importing specific distro variables] ************************************************************************************************************
ok: [localhost] => (item=/root/.ansible/roles/vdo/vars/CentOS.yml)

TASK [vdo : Install the VDO packages] ***********************************************************************************************************************
ok: [localhost]

TASK [vdo : Create the volume group] ************************************************************************************************************************
ok: [localhost]

TASK [vdo : Check if the path exists] ***********************************************************************************************************************
ok: [localhost]

TASK [vdo : Setup VDO pool & LV] ****************************************************************************************************************************
skipping: [localhost]

TASK [vdo : Manage VDO storage pool] ************************************************************************************************************************
changed: [localhost]

TASK [vdo : Manage VDO thin LV] *****************************************************************************************************************************
ok: [localhost]

TASK [vdo : Create vdo mount point] *************************************************************************************************************************
ok: [localhost]

TASK [vdo : Format the xfs filesystem] **********************************************************************************************************************
ok: [localhost]

TASK [vdo : Mount the lv] ***********************************************************************************************************************************
ok: [localhost]

PLAY RECAP **************************************************************************************************************************************************
localhost                  : ok=10   changed=1    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0   

[root@n27-0-253 test]# lvs
  LV     VG    Attr       LSize   Pool   Origin Data%  Meta%  Move Log Cpy%Sync Convert
  vdolv  vdovg vwi-aov--- 100.00g vpool0        0.01                                   
  vpool0 vdovg dwi-ao----  30.00g               13.39                                  
```

Now add 50G to the LV:
```
[root@n27-0-253 test]# vi inventory.yml 
[root@n27-0-253 test]# ansible-playbook main.yaml -i inventory.yml 

PLAY [localhost] ********************************************************************************************************************************************

TASK [Gathering Facts] **************************************************************************************************************************************
ok: [localhost]

TASK [vdo : Importing specific distro variables] ************************************************************************************************************
ok: [localhost] => (item=/root/.ansible/roles/vdo/vars/CentOS.yml)

TASK [vdo : Install the VDO packages] ***********************************************************************************************************************
ok: [localhost]

TASK [vdo : Create the volume group] ************************************************************************************************************************
ok: [localhost]

TASK [vdo : Check if the path exists] ***********************************************************************************************************************
ok: [localhost]

TASK [vdo : Setup VDO pool & LV] ****************************************************************************************************************************
skipping: [localhost]

TASK [vdo : Manage VDO storage pool] ************************************************************************************************************************
ok: [localhost]

TASK [vdo : Manage VDO thin LV] *****************************************************************************************************************************
changed: [localhost]

TASK [vdo : Create vdo mount point] *************************************************************************************************************************
ok: [localhost]

TASK [vdo : Format the xfs filesystem] **********************************************************************************************************************
ok: [localhost]

TASK [vdo : Mount the lv] ***********************************************************************************************************************************
ok: [localhost]

PLAY RECAP **************************************************************************************************************************************************
localhost                  : ok=10   changed=1    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0   

[root@n27-0-253 test]# lvs
  LV     VG    Attr       LSize   Pool   Origin Data%  Meta%  Move Log Cpy%Sync Convert
  vdolv  vdovg vwi-aov--- 150.00g vpool0        0.01                                   
  vpool0 vdovg dwi-ao----  30.00g               13.39                                  
[root@n27-0-253 test]# df -h /mnt/test/ 
Filesystem               Size  Used Avail Use% Mounted on
/dev/mapper/vdovg-vdolv  150G  1.1G  149G   1% /mnt/test
```

Working on the fstype variables now

GregSutcliffe committed:
Make fs configurable

GregSutcliffe commented:
Filesystem is now configurable. I think that just leaves how to handle Zabbix - I can go research what the other roles do, but I willl need your help to actually check it works in Zabbix, @arrfab :)

GregSutcliffe committed:
Use `loop` to allow for multiple VDO LVs

GregSutcliffe commented:
@arrfab added the looping for multiple VDO LVs. Currently assumes a single VG, but I think that's ok, we can easily extend that.

Just Zabbix to go, but a review of the code so far would be great :+1:

GregSutcliffe committed:
Add Zabbix monitoring for VDO physical pools

GregSutcliffe commented:
OK, I've taken a shot at the Zabbix part, based on other roles like "https://github.com/CentOS/ansible-role-mirror/blob/master/tasks/monitoring.yml"

Obviously I don't have a zabbix to hook it up to yet, so I hacked the template to just echo what it *would* do:
```
bash ./zabbix-check-vpool-space.sh 
zabbix_sender -c /etc/zabbix/zabbix_agentd.conf -k vdo.vpool0.percent -o '1741176281 - 30%' >/dev/null
zabbix_sender -c /etc/zabbix/zabbix_agentd.conf -k vdo.vpool0.size -o '1741176281 - 3.2G' >/dev/null
zabbix_sender -c /etc/zabbix/zabbix_agentd.conf -k vdo.vpool1.percent -o '1741176281 - 20%' >/dev/null
zabbix_sender -c /etc/zabbix/zabbix_agentd.conf -k vdo.vpool1.size -o '1741176281 - 4.3G' >/dev/null
```

Which seems about right?

arrfab committed:
Starting with empty list for variables

Signed-off-by: Fabian Arrotin <arrfab@centos.org>

GregSutcliffe committed:
Improve when checks for VDO management

arrfab committed:
pairing review commits

Signed-off-by: Fabian Arrotin <arrfab@centos.org>

GregSutcliffe committed:
Improve idempotence and when-checks on vdo_pools

GregSutcliffe committed:
Add LLD script, cron ; fix Zabbix templates

arrfab commented:
LGTM , tested together on staging instance first . merging to staging and main branch
